### DISCUSSION TITLE
Loop over fields

### DISCUSSION THREAD

**Discussion by morellik** [dflib/dflib#282] [Q&A] [answered]
Dear,

I'm starting to use dflib for my project.   At the end of the various processes, I should cycle through all the fields in the dataframe in order to write the data into a formatted file (not csv).

What is the way to do this?

Thanks

**Comment by morellik** [comment]
For the moment I use the following loop, converting the columnDataFrame objects to string:
 for(RowProxy columnDataFrame: df.rows().select()) {

}

Is there a more common way?

**Comment by andrus** [comment, marked as answer]
Hi Enrico,
DataFrame implements `Iterable<RowProxy>`, so you can iterate over its rows directly (so no need to call `rows().select()`), and within that iterator, you can iterate over columns, either by name or by position. Here are some examples:

```java
// using "forEach"
df.forEach(r -> ...);
```

```java
// using a "for" loop, and iterating over columns by position
int w = df.width();
for(RowProxy rp : df) {
   for(int i = 0; i < w; i++) {
      rp.get(i);
   }
} 
```

```java
// using a "for" loop, and iterating over columns by name
int w = df.width();
for(RowProxy rp : df) {
   for(String column : df.getColumnIndex()) {
      rp.get(column);
   }
} 
```

**Reply by morellik** [reply]
Thank you very much!

**Reply by morellik** [reply]
The fields of each line remain objects to be converted into strings before being written to the file. Right?

**Reply by andrus** [reply]
Of course. The iterator doesn't do any type conversion. It gives you the values that were stored in the DataFrame
